### ISSUE TITLE
Generics g.Map performance

### ISSUE THREAD

**Issue by cezarguimaraes** [maragudk/gomponents#145]
Hi, thank you for the library!

I have been using gomponents to render views of large custom kubernetes resources and noticed that g.Map can be quite inefficient in these scenarios. The issue is that the generics version of g.Map copies every element upon calling the user's callback, which can be quite costly.

I wrote a benchmark to evaluate the performance impact of g.Map against a similar implementation that does not copy elements:

```
goos: linux
goarch: amd64
pkg: github.com/cezarguimaraes/bla
cpu: 12th Gen Intel(R) Core(TM) i7-12700KF
Benchmark_MapRef_BigStruct10x-20            	   46530	     25108 ns/op	  100135 B/op	    1012 allocs/op
Benchmark_MapRef_BigStruct-20               	   47163	     24908 ns/op	   98586 B/op	    1012 allocs/op
Benchmark_gMap_BigStruct10x-20              	     237	   4806597 ns/op	  435981 B/op	    1012 allocs/op
Benchmark_gMap_BigStruct-20                 	    6391	    210963 ns/op	   99668 B/op	    1012 allocs/op
PASS
ok  	github.com/cezarguimaraes/bla	13.882s
```

Notice how `g.Map` performance gets worse when mapping over large elements (8KB and 80KB per element). In particular, a no-copy version of `g.Map` performed ~200 times better for lists of 80KB elements and ~8 times better for 8KB elements.

> Benchmarking code (which also compares performance when preallocating the []g.Node result) available here https://github.com/cezarguimaraes/gomponents_map_bench
Pre allocating the resulting []g.Node also leads to ~25% better performance for ptr Map versions.

It's also worth noting that the previous non-generics version does not have this issue.

---

Another issue with the copying version of g.Map is that it is not suitable for structs that should not be copied, i.e any struct containing a [sync](https://pkg.go.dev/sync) object such as a Mutex, or user-defined values marked [noCopy](https://github.com/golang/go/issues/8005#issuecomment-190753527). Attempting to `g.Map` over these values can lead to subtle bugs, and will cause `go vet` to fail:

```
./main.go:59:42: func passes lock by value: github.com/cezarguimaraes/bla.NoCopyElement contains sync.Mutex
```

---

To conclude, I don't think the library can choose the best performing version for the user while maintaining type safety. However we could support both by either:

1. Adding a `func MapRef func(ts []T, cb(*T) g.Node)` function that doesn't copy the elements
2. Extending `g.Map` to support both pointer and value callbacks:
```go
// type appoximations (~) are deliberately ignored
// so Map can be implemented without reflection
type NodeMapper[T any] interface {
	func(T) g.Node | func(*T) g.Node
}

func Map2[Y NodeMapper[T], T any](ts []T, cb Y) []g.Node {
	nodes := make([]g.Node, 0, len(ts))
	switch f := interface{}(cb).(type) {
	case func(T) g.Node:
		for _, t := range ts {
			nodes = append(nodes, f(t))
		}
		return nodes
	case func(*T) g.Node:
		for i := range ts {
			// &ts[i] is also deliberate as calling cb(&t)
			// destroys performance, possibly due to `t` escaping
			// to the heap
			nodes = append(nodes, f(&ts[i]))
		}
		return nodes
	}
	panic("unreachable")
}

```

In either case, I think there should also be a warning on `Map`  docs regarding poor performance when copying large objects.

**Comment by markuswustenberg** [comment]
Hi @cezarguimaraes, thank you for the detailed issue! You raise some interesting points. I'm currently on paternity leave and will get back to this after. Stay tuned. 😊

**Comment by YuukanOO** [comment]
To make things faster, you can also avoid the `append` call:

```go
func MapPreAlloc[T any](ts []T, cb func(T) g.Node) []g.Node {
	nodes := make([]g.Node, len(ts))
	for i := range ts {
		nodes[i] = cb(ts[i])
	}
	return nodes
}
```

I don't thing there's nothing wrong with the generic `Map`. If you want to avoid the copy of your objects, just hold an array of pointers instead and `Map` should perform just fine on it.

**Comment by cezarguimaraes** [comment]
> If you want to avoid the copy of your objects, just hold an array of pointers in the first place instead and Map should perform just fine on it

It's valid point but in my case and in some common scenarios one does not control the type when rendering a view over:
- imported API types (in my case, custom k8s objects)
- generated types such as protobuf messages

As such one would need to maintain an almost identical type to these 3rd party types. If that is maintenance cost one is willing to pay, this workaround still requires creating N pointers on possibly every render, creating extra unnecessary workload for the GC.

At the very least Map godoc should warn users of the implicit copy.

**Comment by YuukanOO** [comment]
> in some common scenarios one does not control the type when rendering a view over

In those cases, going with a new `MapRef` function seems like a good idea to me to make things more explicit.

**Comment by markuswustenberg** [comment]
Hi @cezarguimaraes . Thank you for your patience. 😊

Could you give an example where using a large struct type you don't control cannot be passed to `Map` as a slice of pointers to instances of that type?

For me, it's clear from the method signature that structs will be copied if you don't pass a pointer type, since that's how all functions in Go work. In your opinion, how does this function differ? (I'm not opposed to the idea of documenting this, I'm just trying to understand your thought process.)

**Comment by cezarguimaraes** [comment]
Hi @markuswustenberg ,

> Could you give an example where using a large struct type you don't control cannot be passed to Map as a slice of pointers to instances of that type?

One can always map over the original slice to a slice of pointers, it's just worse DX in my opinion to have to do so everytime I have to call map with large structs (or `noCopy` elements in the structs). It is also less efficient as we might be creating thousands of new objects for go's GC to take care of every render.

The example I mentioned are types like this: https://github.com/tektoncd/pipeline/blob/fd4cb4621960715439bf0c0757f5be9cad390568/pkg/apis/pipeline/v1/pipelinerun_types.go#L606. `List` types that are autogenerated for kubernetes resources and returned by kubeapi calls.

> For me, it's clear from the method signature that structs will be copied if you don't pass a pointer type, since that's how all functions in Go work. In your opinion, how does this function differ?

I admit it should be apparent for anyone familiar with Go that a copy happens in `Map`. I just think it shouldn't and it couldn't in the non-generic `Map` implementation: https://github.com/maragudk/gomponents/blob/d81de8319f83a4f62ae33f9ada8da267aa9197fd/gomponents_non_generic.go#L7-L13

**Comment by markuswustenberg** [comment]
@cezarguimaraes ,

I agree that it's worse DX, but I think that's okay since I don't think passing large/uncopyable structs into the view layer is a very common use case. I would rather keep things simple in the library. In your case, you could create helper functions to create your pointers to the large structs and use those everywhere in your view package. You might want to do that anyway, since you're already copying your large structs around between packages, no? Or are you creating them in the view layer?

And yes, the non-generic version worked differently, because the more intuitive generic version wasn't possible then. But as stated above, it's perfectly possible for you to just use pointers in the generic Map implementation, and I think that's definitely the least surprising and most intuitive behaviour.

In conclusion, I'm leaning towards not making any changes here.
